mPowerSDK/Dashboard/APHWebviewViewController.m

```diff
@@ -0,0 +1,261 @@
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(timeInterval * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+            [weakSelf savePDFFromWebView:webView];
+        });
+    }
+}
+
+- (void)savePDFFromWebView:(UIWebView*)webView {
+    if (self.printing) {
+        return;
+    }
+    self.printing = YES;
+    [self trySavePDFFromWebView:webView retryCount:0];
+}
+
+- (void)trySavePDFFromWebView:(UIWebView*)webView retryCount:(NSUInteger)retryCount {
+    
+    if (self.isCancelled || (self.pdfURL != nil)) {
+        return;
+    }
+    
+    NSURL *savedPDF = nil;
+    @try {
+        
+        // setup the temporary file
+        NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
+        [dateFormatter setDateFormat:@"yyyy-MM-dd"];
+        NSString *fileName = [NSString stringWithFormat:@"MonthlyReport_%@.pdf", [dateFormatter stringFromDate:[NSDate date]]];
+        NSString *tempDir = NSTemporaryDirectory();
+        NSString *filepath = [tempDir stringByAppendingPathComponent:fileName];
+        if (![[NSFileManager defaultManager] fileExistsAtPath:tempDir]) {
+            NSError *error;
+            [[NSFileManager defaultManager] createDirectoryAtPath:tempDir withIntermediateDirectories:YES attributes:nil error:&error];
+        }
+        savedPDF = [NSURL fileURLWithPath:filepath];
+        
+        // setup the renderer
+        SBAPDFPrintPageRenderer *renderer = [[SBAPDFPrintPageRenderer alloc] init];
+        [renderer addPrintFormatter:webView.viewPrintFormatter startingAtPageAtIndex:0];
+        
+        // Draw to a file
+        CGSize pageSize = [renderer pageSize];
+        CGRect pageRect = CGRectMake(0, 0, pageSize.width, pageSize.height);
+        
+        UIGraphicsBeginPDFContextToFile(filepath, pageRect, nil);
+        
+        NSInteger pages = [renderer numberOfPages];
+        [renderer prepareForDrawingPages:NSMakeRange(0, pages)];
+        
+        for (NSInteger i = 0; i < pages && !self.isCancelled; i++) {
+            UIGraphicsBeginPDFPage();
+            [renderer drawPageAtIndex:i inRect:renderer.printableRect];
+        }
+        
+        UIGraphicsEndPDFContext();
+        
+    }
+    @catch (NSException *exception) {
+        NSLog(@"Error printing: %@", exception);
+        
+        // retry once
+        if (retryCount < 1) {
+            __weak typeof(self) weakSelf = self;
+            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+                [weakSelf trySavePDFFromWebView:webView retryCount:retryCount + 1];
+            });
+        }
+        return;
+    }
+    
+    // Cleanup
+    self.pdfURL = savedPDF;
+    self.pdfWebView.delegate = nil;
+    self.pdfWebView = nil;
+    if (self.isCancelled) {
+        [self deleteTempPDFFile];
+    }
+    self.printing = NO;
+    self.shareButton.enabled = YES;
+}
+
+- (void)deleteTempPDFFile {
+    @synchronized(self) {
```

Comment:
Can you add a TODO item to use a serial queue for this when we have time to revisit? @synchronized works here for now but I'm not a fan of it in general; it's heavyweight, blocking, and prone to deadlocking if you're not careful about call trees.